src/mappings/bank/supply.ts

```diff
@@ -0,0 +1,64 @@
+import type {
+  Coin,
+  CosmosBlock,
+} from "@subql/types-cosmos";
+import type { QueryTotalSupplyResponse } from "cosmjs-types/cosmos/bank/v1beta1/query";
+import { Supply } from "../../types";
+import { stringify } from "../utils";
+
+export const getSupplyId = function(supply: Coin, block: CosmosBlock): string {
+  return `${supply.denom}@${block.block.id}`;
+};
+
+export const getSupplyRecord = function(supply: Coin, block: CosmosBlock): Supply {
+  return Supply.create({
+    id: getSupplyId(supply, block),
+    denom: supply.denom,
+    amount: BigInt(supply.amount),
+    blockId: block.block.id,
+  });
+};
+
+export async function queryTotalSupply(): Promise<Coin[]> {
+  const finalSupply: Coin[] = [];
+  let paginationKey: Uint8Array | undefined;
+
+  try {
+    // Here we force the use of a private property, breaking typescript limitation, due to the need of call a total supply
+    // rpc query of cosmosjs that is not exposed on the implemented client by SubQuery team.
+    // To avoid this, we need to move to implement our own rpc client and also use `unsafe` parameter which I prefer to avoid.
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const queryClient = api.forceGetQueryClient();
+
+    // Initial call to get the first set of results
+    const initialResponse: QueryTotalSupplyResponse = await queryClient.bank.totalSupply() as unknown as QueryTotalSupplyResponse;
+    logger.debug(`[handleTotalSupply]: initialResponse=${stringify(initialResponse, undefined, 2)}`);
+    finalSupply.push(...initialResponse.supply);
+    paginationKey = initialResponse.pagination?.nextKey;
+
+    // Continue fetching if there is a nextKey
+    while (paginationKey && paginationKey.length > 0) {
+      logger.debug(`[handleTotalSupply]: loading more supply pages pagination.nextKey=${JSON.stringify(paginationKey, undefined, 2)}`);
+      const response = await queryClient.bank.totalSupply(paginationKey);
+      finalSupply.push(...response.supply);
+      paginationKey = response.pagination?.nextKey;
+    }
+    logger.debug(`[handleTotalSupply]: all_total_supply=${JSON.stringify(finalSupply, undefined, 2)}`);
+  } catch (error) {
+    logger.error(`[handleTotalSupply] errored: ${error}`);
+  }
+
+  return finalSupply;
+}
+
+export async function _handleSupply(block: CosmosBlock): Promise<void> {
```

Comment:
Wdyt about only adding a supply if the amount has changed since the last supply entity that was created. This will emulate the behavior we would get after switching supply updates to react to claim settlement events.

Reply:
We discuss this yesterday with @Alann27 but then when we need to query like, `get the supply for every block from block X to Y` to query is not doable.

Reply:
After our discussion I update the code to ensure that we get a new record of Supply only under supply changes are received and every block has a relation to the right supply record to been able to accomplish the needs of the explorer.